src/VisualStudio/Core/Def/Implementation/ProjectSystem/FileChangeWatcher.cs

```diff
@@ -0,0 +1,275 @@
+                    _activeFileWatchingTokens.Add(token);
+                }
+
+                _fileChangeWatcher.EnqueueWork(service =>
+                {
+                    uint cookie;
+                    ErrorHandler.ThrowOnFailure(service.AdviseFileChange(filePath, (uint)(_VSFILECHANGEFLAGS.VSFILECHG_Size | _VSFILECHANGEFLAGS.VSFILECHG_Time), this, out cookie));
+
+                    lock (_gate)
+                    {
+                        token.Cookie = cookie;
+                    }
+                });
+
+                return token;
+            }
+
+            public void StopWatchingFile(IFileWatchingToken token)
+            {
+                var typedToken = token as FileWatchingToken;
+
+                if (typedToken == null)
+                {
+                    throw new ArgumentException("The token passed did not originate from this service.", nameof(token));
+                }
+
+                if (typedToken == _noOpFileWatchingToken)
+                {
+                    // This file never required a direct file watch, our main subscription covered it.
+                    return;
+                }
+
+                lock (_gate)
+                {
+                    if (!_activeFileWatchingTokens.Remove(typedToken))
+                    {
+                        throw new InvalidOperationException("This token was no longer being watched.");
+                    }
+                }
+
+                _fileChangeWatcher.EnqueueWork(service =>
+                    ErrorHandler.ThrowOnFailure(service.UnadviseFileChange(typedToken.Cookie.Value)));
+            }
+
+            public event EventHandler<string> FileChanged;
+
+            int IVsFreeThreadedFileChangeEvents.FilesChanged(uint cChanges, string[] rgpszFile, uint[] rggrfChange)
+            {
+                for (int i = 0; i < cChanges; i++)
+                {
+                    FileChanged?.Invoke(this, rgpszFile[i]);
+                }
+
+                return VSConstants.S_OK;
+            }
+
+            int IVsFreeThreadedFileChangeEvents.DirectoryChanged(string pszDirectory)
+            {
+                return VSConstants.E_NOTIMPL;
+            }
+
+            int IVsFreeThreadedFileChangeEvents.DirectoryChangedEx(string pszDirectory, string pszFile)
+            {
+                FileChanged?.Invoke(this, pszFile);
+
+                return VSConstants.S_OK;
+            }
+
+            int IVsFileChangeEvents.FilesChanged(uint cChanges, string[] rgpszFile, uint[] rggrfChange)
+            {
+                for (int i = 0; i < cChanges; i++)
+                {
+                    FileChanged?.Invoke(this, rgpszFile[i]);
+                }
+
+                return VSConstants.S_OK;
+            }
+
+            int IVsFileChangeEvents.DirectoryChanged(string pszDirectory)
+            {
+                return VSConstants.E_NOTIMPL;
+            }
+
+            public class FileWatchingToken : IFileWatchingToken
```

Comment:
>class [](start = 19, length = 5)

Do we need object identity here or can we use the value of Cookie as the identity? If we don't need identity then make this struct, Cookie field private and remove IFileWatchingToken.

Reply:
We need identity, since in the current implementation this is updated asynchronously on a separate thread.